File: src/core/execution/engine.py
"""
Execution Engine for Obelisk Core
Executes node-based workflows (similar to ComfyUI)
"""
from typing import Dict, Any, List, Optional, Set
from collections import defaultdict, deque
from ..types import NodeGraph, NodeID, ConnectionData, NodeData, GraphExecutionResult, NodeExecutionResult
from .node_base import BaseNode, ExecutionContext
from .node_registry import get_node_class
from ..container import ServiceContainer
from ...utils.logger import get_logger
import time

logger = get_logger(__name__)


class CycleError(ValueError):
    """Raised when a cycle is detected in the workflow graph"""
    pass


class ExecutionEngine:
    """
    Executes JSON workflow graphs
    
    Features:
    - Topological sort for execution order
    - Dependency resolution
    - Connection handling between nodes
    - State management across execution
    """
    
    def __init__(self, container: Optional[ServiceContainer] = None):
        """
        Initialize execution engine
        
        Args:
            container: Optional ServiceContainer (for backward compatibility, nodes should be self-contained)
        """
        self.container = container
    
    def execute(self, workflow: NodeGraph, context_variables: Optional[Dict[str, Any]] = None) -> GraphExecutionResult:
        """
        Execute a workflow graph
        
        Args:
            workflow: NodeGraph definition (from JSON)
            context_variables: Runtime variables (e.g., {"user_id": "cli_user", "user_query": "Hello"})
            
        Returns:
            GraphExecutionResult with execution results
        """
        start_time = time.time()
        context_variables = context_variables or {}
        
        logger.info(f"Executing workflow: {workflow.get('name', workflow.get('id', 'unknown'))}")
        
        # Validate graph
        if not self.validate_graph(workflow):
            return GraphExecutionResult(
                graph_id=workflow.get('id', 'unknown'),
                success=False,
                node_results=[],
                final_outputs={},
                error="Graph validation failed",
                total_execution_time=time.time() - start_time
            )
        
        # Build node instances
        nodes = self._build_nodes(workflow)
        
        # Resolve execution order (topological sort)
        try:
            execution_order = self.resolve_execution_order(workflow, nodes)
        except CycleError as e:
            return GraphExecutionResult(
                graph_id=workflow.get('id', 'unknown'),
                success=False,
                node_results=[],
                final_outputs={},
                error=f"Cycle detected in workflow graph: {str(e)}",
                total_execution_time=time.time() - start_time
            )
        
        # Create execution context
        context = ExecutionContext(
            container=self.container,
            variables=context_variables,
            node_outputs={}
        )
        
        # Execute nodes in order
        node_results: List[NodeExecutionResult] = []
        errors: List[str] = []
        
        for node_id in execution_order:
            node = nodes[node_id]
            node_start_time = time.time()
            
            try:
                # Resolve inputs from connections
                resolved_inputs = self._resolve_node_inputs(node, workflow, context)
                
                # Update node inputs with resolved values
                original_inputs = node.inputs.copy()
                node.inputs.update(resolved_inputs)
                
                # Execute node
                outputs = node.execute(context)
                
                # Store outputs in context
                context.node_outputs[node_id] = outputs
                
                # Restore original inputs
                node.inputs = original_inputs
                
                execution_time = time.time() - node_start_time
                
                node_results.append(NodeExecutionResult(
                    node_id=node_id,
                    success=True,
                    outputs=outputs,
                    execution_time=execution_time
                ))
                
                logger.debug(f"Node {node_id} ({node.node_type}) executed successfully in {execution_time:.3f}s")
                
            except Exception as e:
                execution_time = time.time() - node_start_time
                error_msg = f"Node {node_id} ({node.node_type}) failed: {str(e)}"
                errors.append(error_msg)
                logger.error(error_msg, exc_info=True)
                
                node_results.append(NodeExecutionResult(
                    node_id=node_id,
                    success=False,
                    outputs={},
                    error=str(e),
                    execution_time=execution_time
                ))
                
                # Stop execution on error (can be made configurable)
                break
        
        # Collect final outputs (from output_text nodes)
        final_outputs = self._collect_final_outputs(workflow, context)
        
        total_time = time.time() - start_time
        success = len(errors) == 0
        
        logger.info(f"Workflow execution {'succeeded' if success else 'failed'} in {total_time:.3f}s")
        
        return GraphExecutionResult(
            graph_id=workflow.get('id', 'unknown'),
            success=success,
            node_results=node_results,
            final_outputs=final_outputs,
            error="; ".join(errors) if errors else None,
            total_execution_time=total_time,
            execution_order=execution_order  # Include execution order for frontend highlighting
        )
    
    def validate_graph(self, workflow: NodeGraph) -> bool:
        """
        Validate workflow graph structure
        
        Args:
            workflow: NodeGraph to validate
            
        Returns:
            True if valid, False otherwise
        """
        if 'nodes' not in workflow or not workflow['nodes']:
            logger.error("Workflow has no nodes")
            return False
        
        if 'connections' not in workflow:
            workflow['connections'] = []
        
        # Check all connections reference valid nodes
        node_ids = {node['id'] for node in workflow['nodes']}
        
        for conn in workflow['connections']:
            if conn.get('source_node') not in node_ids:
                logger.error(f"Connection references invalid source node: {conn.get('source_node')}")
                return False
            if conn.get('target_node') not in node_ids:
                logger.error(f"Connection references invalid target node: {conn.get('target_node')}")
                return False
        
        # Check all node types are registered
        for node in workflow['nodes']:
            node_type = node.get('type')
            if not node_type:
                logger.error(f"Node {node.get('id')} has no type")
                return False
            
            if not get_node_class(node_type):
                logger.error(f"Unknown node type: {node_type}")
                return False
        
        return True
    
    def resolve_execution_order(self, workflow: NodeGraph, nodes: Dict[NodeID, BaseNode]) -> List[NodeID]:
        """
        Resolve execution order using topological sort
        
        Args:
            workflow: NodeGraph definition
            nodes: Dictionary of node_id -> node instance
            
        Returns:
            List of node IDs in execution order
        """
        # Build dependency graph
        # dependencies[node_id] = set of node IDs that must execute before this node
        dependencies: Dict[NodeID, Set[NodeID]] = defaultdict(set)
        in_degree: Dict[NodeID, int] = defaultdict(int)
        
        # Initialize all nodes
        for node_id in nodes.keys():
            in_degree[node_id] = 0
        
        # Process connections to build dependency graph
        connections = workflow.get('connections', [])
        for conn in connections:
            source_id = conn['source_node']
            target_id = conn['target_node']
            
            if source_id not in dependencies[target_id]:
                dependencies[target_id].add(source_id)
                in_degree[target_id] += 1
        
        # Topological sort (Kahn's algorithm)
        queue = deque([node_id for node_id, degree in in_degree.items() if degree == 0])
        execution_order: List[NodeID] = []
        
        while queue:
            node_id = queue.popleft()
            execution_order.append(node_id)
            
            # Find nodes that depend on this node
            for target_id, deps in dependencies.items():
                if node_id in deps:
                    in_degree[target_id] -= 1
                    if in_degree[target_id] == 0:
                        queue.append(target_id)
        
        # Check for cycles (shouldn't happen in DAG, but fail fast if detected)
        if len(execution_order) != len(nodes):
            # Find nodes that weren't included in execution order (part of cycle)
            executed_node_ids = set(execution_order)
            cycle_nodes = [node_id for node_id in nodes.keys() if node_id not in executed_node_ids]
            raise CycleError(
                f"Cycle detected in workflow graph: {len(execution_order)}/{len(nodes)} nodes in execution order. "
                f"Nodes involved in cycle: {cycle_nodes}"
            )
        
        return execution_order
    
    def _build_nodes(self, workflow: NodeGraph) -> Dict[NodeID, BaseNode]:
        """
        Build node instances from workflow
        
        Args:
            workflow: NodeGraph definition
            
        Returns:
            Dictionary of node_id -> node instance
        """
        nodes: Dict[NodeID, BaseNode] = {}
        
        # First pass: create all nodes (without workflow/all_nodes)
        for node_data in workflow['nodes']:
            node_id = node_data['id']
            node_type = node_data['type']
            
            node_class = get_node_class(node_type)
            if not node_class:
                raise ValueError(f"Unknown node type: {node_type}")
            
            nodes[node_id] = node_class(node_id, node_data)
        
        # Second pass: setup all nodes (now they can discover other nodes)
        for node in nodes.values():
            node._setup(workflow, nodes)
        
        return nodes
    
    def _resolve_node_inputs(self, node: BaseNode, workflow: NodeGraph, context: ExecutionContext) -> Dict[str, Any]:
        """
        Resolve node inputs from connections and context variables
        
        Args:
            node: Node to resolve inputs for
            workflow: Workflow definition
            context: Execution context
            
        Returns:
            Dictionary of resolved input values
        """
        resolved = {}
        connections = workflow.get('connections', [])
        
        # Find all connections targeting this node
        for conn in connections:
            if conn['target_node'] != node.node_id:
                continue
            
            source_id = conn['source_node']
            source_output = conn.get('source_output', 'default')
            target_input = conn.get('target_input', 'default')
            
            # Get output from source node
            if source_id in context.node_outputs:
                source_outputs = context.node_outputs[source_id]
                if source_output in source_outputs:
                    resolved[target_input] = source_outputs[source_output]
        
        # Also resolve template variables in direct inputs
        for input_name, input_value in node.inputs.items():
            if input_name not in resolved:  # Don't override connections
                if isinstance(input_value, str) and input_value.startswith('{{') and input_value.endswith('}}'):
                    var_name = input_value[2:-2].strip()
                    # Only resolve if variable exists in context (don't overwrite with None)
                    if var_name in context.variables:
                        resolved[input_name] = context.variables[var_name]
                    # If variable doesn't exist, leave unresolved so get_input_value() can use defaults
                else:
                    resolved[input_name] = input_value
        
        return resolved
    
    def _collect_final_outputs(self, workflow: NodeGraph, context: ExecutionContext) -> Dict[str, Any]:
        """
        Collect final outputs from output_text nodes
        
        Args:
            workflow: Workflow definition
            context: Execution context
            
        Returns:
            Dictionary of final outputs
        """
        outputs = {}
        
        for node_data in workflow['nodes']:
            if node_data['type'] == 'output_text':
                node_id = node_data['id']
                if node_id in context.node_outputs:
                    outputs.update(context.node_outputs[node_id])
        
        return outputs

File: ts/src/core/execution/engine.ts
/**
 * Execution Engine – topological sort + DAG execution.
 * Mirrors Python src/core/execution/engine.py
 *
 * 1. Validates workflow graph structure
 * 2. Builds a DAG and computes topological execution order
 * 3. Instantiates node classes via the registry
 * 4. Executes nodes in order, propagating outputs via connections
 *    – Stops on first error (matches Python behaviour)
 */
import {
  NodeID,
  WorkflowData,
  ConnectionData,
  GraphExecutionResult,
  normalizeWorkflowConnections,
} from "../types";
import { BaseNode, ExecutionContext } from "./nodeBase";
import { getNodeClass, registerAllNodes } from "./nodeRegistry";
import { getLogger, abbrevPathForLog, sanitizeForLog } from "../../utils/logger";

const logger = getLogger("engine");
const SKIP_DEBUG_KEYS = new Set(["storage_instance"]);

export class CycleError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CycleError";
  }
}

export class ExecutionEngine {
  constructor() {
    // Ensure all node types are registered
    registerAllNodes();
  }

  // ── Main entry point ────────────────────────────────────────────────

  /**
   * Execute a workflow graph.
   *
   * @param workflow  The workflow JSON (nodes + connections)
   * @param contextVariables  Variables injected into execution (user_query, etc.)
   * @param initialNodeOutputs  Pre-seeded node outputs (e.g. from autonomous nodes in the runner)
   * @returns Execution result with per-node outputs
   */
  async execute(
    workflow: WorkflowData,
    contextVariables: Record<string, unknown> = {},
    initialNodeOutputs: Record<NodeID, Record<string, unknown>> = {}
  ): Promise<GraphExecutionResult> {
    const startTime = Date.now();

    // Normalise connections once at the boundary — all downstream code can
    // rely on source_node / target_node being present.
    normalizeWorkflowConnections(workflow);

    logger.info(
      `Executing workflow: ${workflow.name ?? workflow.id ?? "unknown"}`
    );

    try {
      // 1. Validate graph (allow external sources if their outputs are provided)
      const externalSources = new Set(Object.keys(initialNodeOutputs));
      if (!this.validateGraph(workflow, externalSources)) {
        return {
          graphId: workflow.id ?? "unknown",
          success: false,
          nodeResults: [],
          finalOutputs: {},
          error: "Graph validation failed",
          totalExecutionTime: Date.now() - startTime,
        };
      }

      // 2. Build node instances
      const nodeMap = this.buildNodeMap(workflow);
      if (!nodeMap.size) {
        return {
          graphId: workflow.id ?? "unknown",
          success: true,
          nodeResults: [],
          finalOutputs: {},
          executionOrder: [],
          totalExecutionTime: 0,
        };
      }

      // 3. Second pass: setup all nodes (allows cross-node discovery)
      for (const node of nodeMap.values()) {
        node._setup(workflow, nodeMap);
      }

      // 4. Resolve execution order (topological sort)
      let order: NodeID[];
      try {
        order = this.topologicalSort(nodeMap, workflow.connections);
      } catch (err) {
        if (err instanceof CycleError) {
          return {
            graphId: workflow.id ?? "unknown",
            success: false,
            nodeResults: [],
            finalOutputs: {},
            error: `Cycle detected in workflow graph: ${err.message}`,
            totalExecutionTime: Date.now() - startTime,
          };
        }
        throw err;
      }

      // 5. Create execution context with pre-seeded outputs
      const context: ExecutionContext = {
        variables: { ...contextVariables },
        nodeOutputs: { ...initialNodeOutputs }, // copy to avoid mutation
      };

      // 6. Execute nodes in order
      const nodeResults: GraphExecutionResult["nodeResults"] = [];
      const errors: string[] = [];

      for (const nodeId of order) {
        const node = nodeMap.get(nodeId);
        if (!node) continue;

        const nodeStart = Date.now();
        try {
          // Resolve inputs from connections (mirrors Python _resolve_node_inputs)
          const resolvedInputs = this.resolveNodeInputs(
            node,
            workflow,
            context
          );

          // Save original inputs, apply resolved values
          const originalInputs = { ...node.inputs };
          Object.assign(node.inputs, resolvedInputs);

          // Execute node
          const outputs = await node.execute(context);

          // Store outputs in context
          context.nodeOutputs[nodeId] = outputs;

          // Restore original inputs (prevents side-effects across ticks)
          node.inputs = originalInputs;

          const nodeExecTime = Date.now() - nodeStart;
          nodeResults.push({
            nodeId,
            success: true,
            outputs,
            executionTime: nodeExecTime,
          });

          // Log per-node execution at INFO level (truncated summary; paths abbreviated with ~)
          const outputKeys = Object.keys(outputs);
          const outputSummary = outputKeys
            .map((k) => {
              const v = outputs[k];
              if (v === null || v === undefined) return `${k}=null`;
              if (typeof v === "string") {
                const s = abbrevPathForLog(v);
                return `${k}="${s.length > 60 ? s.slice(0, 60) + "..." : s}"`;
              }
              if (typeof v === "boolean" || typeof v === "number") return `${k}=${v}`;
              return `${k}=<${typeof v}>`;
            })
            .join(", ");
          logger.info(
            `  Node ${nodeId} (${node.nodeType}) → ${nodeExecTime}ms [${outputSummary}]`
          );

          // DEBUG: log full outputs when OBELISK_CORE_DEBUG=true; cap size to avoid huge logs (e.g. full clanker state)
          const MAX_DEBUG_PAYLOAD = 2000;
          for (const k of outputKeys) {
            if (SKIP_DEBUG_KEYS.has(k)) continue;
            const v = outputs[k];
            if (typeof v === "string") {
              const s = abbrevPathForLog(v);
              const fullInference = node.nodeType === "inference" && (k === "query" || k === "response");
              if (fullInference || s.length <= MAX_DEBUG_PAYLOAD) {
                logger.debug(`  [${nodeId}] FULL ${k} (${s.length} chars):\n${s}`);
              } else {
                logger.debug(`  [${nodeId}] ${k}: string ${s.length} chars (truncated in debug)`);
              }
            } else if (v !== null && v !== undefined && typeof v === "object") {
              try {
                const sanitized = sanitizeForLog(v);
                const json = JSON.stringify(sanitized, null, 2);
                if (json.length <= MAX_DEBUG_PAYLOAD) {
                  logger.debug(`  [${nodeId}] FULL ${k}:\n${json}`);
                } else {
                  logger.debug(`  [${nodeId}] ${k}: object ${json.length} chars (truncated in debug)`);
                }
              } catch { /* skip non-serialisable */ }
            }
          }

          // Always log model output at INFO so it's visible without DEBUG (for inference + debug Text node)
          const MODEL_OUTPUT_MAX = 1200;
          if (node.nodeType === "inference" && typeof outputs.response === "string") {
            const s = outputs.response;
            logger.info(`  [Model output] ${s.length <= MODEL_OUTPUT_MAX ? s : s.slice(0, MODEL_OUTPUT_MAX) + "..."}`);
          }
          if (node.nodeType === "text" && typeof outputs.text === "string" && outputs.text.length > 0) {
            const s = outputs.text;
            logger.info(`  [Debug text node ${nodeId}] ${s.length <= MODEL_OUTPUT_MAX ? s : s.slice(0, MODEL_OUTPUT_MAX) + "..."}`);
          }
        } catch (err) {
          const errorMsg = err instanceof Error ? err.message : String(err);
          const fullErrMsg = `Node ${nodeId} (${node.nodeType}) failed: ${errorMsg}`;
          errors.push(fullErrMsg);
          logger.error(`  ✗ ${fullErrMsg}`);

          nodeResults.push({
            nodeId,
            success: false,
            outputs: {},
            error: errorMsg,
            executionTime: Date.now() - nodeStart,
          });

          // Stop execution on error (matches Python behaviour)
          break;
        }
      }

      // 7. Collect final outputs (from output_text nodes – matches Python)
      const finalOutputs = this.collectFinalOutputs(workflow, context);

      const overallSuccess = errors.length === 0;

      logger.info(
        `Workflow execution ${overallSuccess ? "succeeded" : "failed"} in ${Date.now() - startTime}ms`
      );

      return {
        graphId: workflow.id,
        success: overallSuccess,
        nodeResults,
        finalOutputs,
        error: errors.length ? errors.join("; ") : undefined,
        executionOrder: order,
        totalExecutionTime: Date.now() - startTime,
      };
    } catch (err) {
      const errorMsg = err instanceof Error ? err.message : String(err);
      logger.error(`Execution failed: ${errorMsg}`);
      return {
        graphId: workflow.id,
        success: false,
        nodeResults: [],
        finalOutputs: {},
        error: errorMsg,
        totalExecutionTime: Date.now() - startTime,
      };
    }
  }

  // ── Graph validation (mirrors Python validate_graph) ────────────────

  /**
   * Validate workflow graph structure.
   *
   * @param workflow  The workflow to validate
   * @param externalSources  Node IDs that are valid sources even if not in
   *   the workflow (used for autonomous nodes whose outputs are passed via
   *   initialNodeOutputs)
   */
  validateGraph(
    workflow: WorkflowData,
    externalSources: Set<NodeID> = new Set()
  ): boolean {
    if (!workflow.nodes || workflow.nodes.length === 0) {
      logger.error("Workflow has no nodes");
      return false;
    }

    if (!workflow.connections) {
      workflow.connections = [];
    }

    // Check all connections reference valid nodes
    const nodeIds = new Set(workflow.nodes.map((n) => String(n.id)));

    for (const conn of workflow.connections) {
      if (
        !nodeIds.has(conn.source_node) &&
        !externalSources.has(conn.source_node)
      ) {
        logger.error(
          `Connection references invalid source node: ${conn.source_node}`
        );
        return false;
      }
      if (!nodeIds.has(conn.target_node)) {
        logger.error(
          `Connection references invalid target node: ${conn.target_node}`
        );
        return false;
      }
    }

    // Check all node types are registered
    for (const node of workflow.nodes) {
      if (!node.type) {
        logger.error(`Node ${node.id} has no type`);
        return false;
      }
      if (!getNodeClass(node.type)) {
        logger.error(`Unknown node type: ${node.type}`);
        return false;
      }
    }

    return true;
  }

  // ── Helpers (public so WorkflowRunner can use for subgraph building) ─

  buildNodeMap(workflow: WorkflowData): Map<NodeID, BaseNode> {
    const map = new Map<NodeID, BaseNode>();
    for (const nodeData of workflow.nodes) {
      const Ctor = getNodeClass(nodeData.type);
      if (!Ctor) {
        logger.error(
          `Unknown node type "${nodeData.type}" for node ${nodeData.id} – skipping`
        );
        continue;
      }
      map.set(nodeData.id, new Ctor(nodeData.id, nodeData));
    }
    return map;
  }

  /**
   * Wire input connections so each node knows where its inputs come from.
   */
  wireConnections(
    nodeMap: Map<NodeID, BaseNode>,
    connections: ConnectionData[]
  ): void {
    for (const conn of connections) {
      const targetNode = nodeMap.get(conn.target_node);
      if (!targetNode) continue;
      if (!targetNode.inputConnections[conn.target_input]) {
        targetNode.inputConnections[conn.target_input] = [];
      }
      targetNode.inputConnections[conn.target_input].push({
        nodeId: conn.source_node,
        outputName: conn.source_output,
      });
    }
  }

  /**
   * Resolve node inputs from connections and context variables.
   * Mirrors Python _resolve_node_inputs.
   */
  resolveNodeInputs(
    node: BaseNode,
    workflow: WorkflowData,
    context: ExecutionContext
  ): Record<string, unknown> {
    const resolved: Record<string, unknown> = {};
    const connections = workflow.connections ?? [];

    // Find all connections targeting this node
    for (const conn of connections) {
      if (conn.target_node !== String(node.nodeId)) continue;

      // Get output from source node
      const sourceOutputs = context.nodeOutputs[conn.source_node];
      if (sourceOutputs) {
        const val = sourceOutputs[conn.source_output];
        if (val !== undefined) {
          resolved[conn.target_input] = val;
        }
      }
    }

    // Also resolve template variables in direct inputs
    for (const [inputName, inputValue] of Object.entries(node.inputs)) {
      if (inputName in resolved) continue; // Don't override connections

      if (
        typeof inputValue === "string" &&
        inputValue.startsWith("{{") &&
        inputValue.endsWith("}}")
      ) {
        const varName = inputValue.slice(2, -2).trim();
        // Check process.env first
        if (varName.startsWith("process.env.")) {
          const envKey = varName.slice("process.env.".length);
          if (process.env[envKey] !== undefined) {
            resolved[inputName] = process.env[envKey];
            logger.debug(
              `[Engine] Resolved env template ${inputName}={{${varName}}} for node ${node.nodeId}`
            );
          } else {
            logger.warning(
              `[Engine] Env template ${inputName}={{${varName}}} not found for node ${node.nodeId}`
            );
          }
        } else if (varName in context.variables) {
          resolved[inputName] = context.variables[varName];
          logger.debug(
            `[Engine] Resolved template variable ${inputName}={{${varName}}} to '${context.variables[varName]}' for node ${node.nodeId}`
          );
        } else {
          // Variable doesn't exist - log warning and leave unresolved
          logger.warning(
            `[Engine] Template variable ${inputName}={{${varName}}} not found in context.variables (available: ${Object.keys(context.variables).join(", ")}) for node ${node.nodeId}`
          );
        }
      } else {
        resolved[inputName] = inputValue;
      }
    }

    return resolved;
  }

  /**
   * Kahn's algorithm for topological sort.
   * Throws CycleError if the graph contains a cycle.
   */
  topologicalSort(
    nodeMap: Map<NodeID, BaseNode>,
    connections: ConnectionData[]
  ): NodeID[] {
    const nodeIds = Array.from(nodeMap.keys());
    const inDegree: Record<NodeID, number> = {};
    const adjacency: Record<NodeID, Set<NodeID>> = {};

    for (const id of nodeIds) {
      inDegree[id] = 0;
      adjacency[id] = new Set();
    }

    for (const conn of connections) {
      if (!nodeMap.has(conn.source_node) || !nodeMap.has(conn.target_node))
        continue;
      // Avoid counting duplicate edges
      if (!adjacency[conn.source_node].has(conn.target_node)) {
        adjacency[conn.source_node].add(conn.target_node);
        inDegree[conn.target_node]++;
      }
    }

    const queue: NodeID[] = nodeIds.filter((id) => inDegree[id] === 0);
    const sorted: NodeID[] = [];

    while (queue.length > 0) {
      const current = queue.shift()!;
      sorted.push(current);
      for (const neighbour of adjacency[current]) {
        inDegree[neighbour]--;
        if (inDegree[neighbour] === 0) {
          queue.push(neighbour);
        }
      }
    }

    if (sorted.length !== nodeIds.length) {
      const executedSet = new Set(sorted);
      const cycleNodes = nodeIds.filter((id) => !executedSet.has(id));
      throw new CycleError(
        `${sorted.length}/${nodeIds.length} nodes in execution order. ` +
          `Nodes involved in cycle: ${cycleNodes.join(", ")}`
      );
    }

    return sorted;
  }

  /**
   * Collect final outputs from output_text nodes.
   * Mirrors Python _collect_final_outputs.
   */
  private collectFinalOutputs(
    workflow: WorkflowData,
    context: ExecutionContext
  ): Record<string, unknown> {
    const outputs: Record<string, unknown> = {};

    for (const nodeData of workflow.nodes) {
      if (nodeData.type === "output_text") {
        const nodeId = nodeData.id;
        const nodeOutputs = context.nodeOutputs[nodeId];
        if (nodeOutputs) {
          Object.assign(outputs, nodeOutputs);
        }
      }
    }

    return outputs;
  }
}
